### PR TITLE
implement type-hints for all our functions

### PR DESCRIPTION
**Pull request by igalic**
as a first step, we'll ad mypy static analyzer to check those.
This should give us extra (offline) guarantees about those types.

### CODE REVIEW

**Comment by igalic** [comment]
this first attempt at #39, is currently failing to bootstrap:

```
libiocage/cli/__init__.py:38: error: Module has no attribute "_verify_python3_env"
libiocage/cli/__init__.py:102: error: Unexpected keyword argument "message" for "version_option"
/home/igalic/.pyenv/versions/3.6.1/lib/mypy/typeshed/third_party/2and3/click/decorators.pyi:174: note: "version_option" defined here
 ǃ  ~/s/t/libiocage  ╍╍ type-hints 
```

**Comment by gronke** [comment]
Do we need to add an entry to requirements.txt?

**Comment by igalic** [comment]
new set of errors:

```
libiocage/lib/helpers.py:52: error: Invalid tuple literal type
libiocage/lib/Jail.py:67: error: Name 'libzfs' is not defined
libiocage/lib/Jail.py:68: error: Invalid type "libiocage.lib.Host"
libiocage/lib/Jail.py:69: error: Invalid type "libiocage.lib.Logger"
libiocage/cli/__init__.py:38: error: Module has no attribute "_verify_python3_env"
libiocage/cli/__init__.py:102: error: Unexpected keyword argument "message" for "version_option"
/home/igalic/.pyenv/versions/3.6.1/lib/mypy/typeshed/third_party/2and3/click/decorators.pyi:174: note: "version_option" defined here
```

**Comment by igalic** [comment]
> @gronke: Do we need to add an entry to requirements.txt?

i don't think so: it's not a run-time requirement, only for test, hence an entry in `test_requires()`, but not in `requirements.txt`

File: libiocage/lib/Jail.py [hunk 1]
```diff
@@ -60,19 +62,24 @@ class Jail:
             release (stored in `zpool/iocage/base/<RELEASE>)
     """
 
-    def __init__(self, data={}, zfs=None, host=None, logger=None, new=False):
+    def __init__(self,
+                 data: Union[str, dict]={},
```
Comment (skarekrow) [inline libiocage/lib/Jail.py]: Why is this a union? It is never a string, but always a dict.

Comment (igalic) [reply libiocage/lib/Jail.py]: because that's exactly what the pydoc comment below says

Comment (skarekrow) [reply libiocage/lib/Jail.py]: lol, coffee. You're right, sorry!

Comment (igalic) [reply libiocage/lib/Jail.py]: yeah, but, we should still figure out if we want to simplify that

as i have done with the execs that would've allowed `Union[str, List[str]]`

File: libiocage/lib/Jail.py [hunk 2]
```diff
@@ -60,19 +62,24 @@ class Jail:
...
+                 zfs: Optional[libzfs.ZFS]=None,
```
Comment (igalic) [inline libiocage/lib/Jail.py]: i can't verify this on linux ☹

```
libiocage/lib/Jail.py:67: error: Name 'libzfs' is not defined
```

File: libiocage/lib/Jail.py [hunk 3]
```diff
@@ -60,19 +62,24 @@ class Jail:
...
+                 host: Optional[libiocage.lib.Host]=None,
+                 logger: Optional[libiocage.lib.Logger]=None,
```
Comment (igalic) [inline libiocage/lib/Jail.py]: for these two, i'm getting

```
libiocage/lib/Jail.py:68: error: Invalid type "libiocage.lib.Host"
libiocage/lib/Jail.py:69: error: Invalid type "libiocage.lib.Logger"
```

File: libiocage/lib/helpers.py [hunk 4]
```diff
@@ -46,11 +48,10 @@ def init_logger(self, logger=None):
         object.__setattr__(self, 'logger', new_logger)
 
 
-def exec(command, logger=None, ignore_error=False):
-    if isinstance(command, str):
-        command = [command]
+def exec(command, logger=None, ignore_error=False) -> (
+        subprocess.Popen, str, str,):
```
Comment (igalic) [inline libiocage/lib/helpers.py]: this is… wrong, apparently:

```
libiocage/lib/helpers.py:52: error: Invalid tuple literal type
```

File: libiocage/lib/Jail.py (hunk 2 above)
Comment (igalic) [reply libiocage/lib/Jail.py]: which means we won't be able to verify this on Travis…

File: libiocage/lib/Jail.py (hunk 3 above)
Comment (igalic) [reply libiocage/lib/Jail.py]: new data! this is a module, not a type

how do i access the actual type, or the class?

Comment (gronke) [reply libiocage/lib/Jail.py]: libiocage.lib.Host.Host 🌮

Comment (igalic) [reply libiocage/lib/Jail.py]: but that's uglt 😅

Comment (gronke) [reply libiocage/lib/Jail.py]: Yeah, I know! The other option is to put everything in one file. Let's quickly forget that.

Yet another solution is to mess with `sys.path`, but @william-gr taught me this is a no-go for libraries and bad practice everywhere else.

I think `lib/__init__.py` is an appropriate solution: The public exposed classes are available as `libiocage.Jail`, `libiocage.Release`, etc when importing the directory (for development) or importing globally installed libiocage.

File: libiocage/lib/Jail.py (hunk 2 above)
Comment (gronke) [reply libiocage/lib/Jail.py]: Indeed that's a problem. Can we provide a Typing file for libzfs?

**Comment by igalic** [comment]
UPDATE!!
after generating types interfaces for the classes missing them (see: https://github.com/freenas/py-libzfs/pull/4 , https://github.com/vstakhov/libucl/pull/168 & https://github.com/foutaise/texttable/pull/22 ), i now have the following set of errors:

```
libiocage/lib/helpers.py:52: error: Invalid tuple literal type
libiocage/lib/Jail.py:138: error: Incompatible types in assignment (expression has type "RCConf", variable has type None)
libiocage/lib/Jail.py:161: error: Incompatible types in assignment (expression has type Type[NullFSBasejailStorage], variable has type Type[ZFSBasejailStorage])
libiocage/lib/Jail.py:538: error: Argument 1 to "append" of "list" has incompatible type "Network"; expected "str"
libiocage/lib/Jail.py:542: error: "str" has no attribute "teardown"
libiocage/lib/Jail.py:730: error: Value of type None is not indexable
libiocage/lib/Jail.py:736: error: Value of type None is not indexable
libiocage/lib/Jail.py:782: error: Incompatible types in assignment (expression has type "str", variable has type None)
libiocage/lib/Jail.py:796: error: "str" has no attribute "mountpoint"
libiocage/lib/Jail.py:852: error: Type[Dict[Any, Any]] has no attribute "__dir__"
libiocage/lib/Distribution.py:110: error: Need type annotation for variable
libiocage/lib/Distribution.py:115: error: Incompatible types in assignment (expression has type List[str], variable has type "str")
libiocage/lib/Distribution.py:128: error: "Distribution" has no attribute "host"
libiocage/cli/__init__.py:38: error: Module has no attribute "_verify_python3_env"
libiocage/cli/__init__.py:103: error: Unexpected keyword argument "message" for "version_option"
/usr/local/lib/mypy/typeshed/third_party/2and3/click/decorators.pyi:174: note: "version_option" defined here
```

File: libiocage/lib/Jail.py [hunk 5]
```diff
@@ -143,7 +152,15 @@ def start(self):
 
         release = self.release
 
-        backend = None
+        backend: Union[
+            libiocage.lib.StandaloneJailStorage.StandaloneJailStorage,
+            libiocage.lib.NullFSBasejailStorage.NullFSBasejailStorage,
+            libiocage.lib.ZFSBasejailStorage.ZFSBasejailStorage,
+            None,
+        ] = None
```
Comment (igalic) [inline libiocage/lib/Jail.py]: I have no idea what i'm doing

File: libiocage/lib/Jail.py [hunk 6]
```diff
@@ -143,7 +152,15 @@ def start(self):
...
+
+        if self.config["basejail_type"] == "standalone":
+            backend = libiocage.lib.StandaloneJailStorage.StandaloneJailStorage
```
Comment (igalic) [inline libiocage/lib/Jail.py]:
```
libiocage/lib/Jail.py:163: error: Incompatible types in assignment (expression has type Type[StandaloneJailStorage], variable has type "Union[StandaloneJailStorage, NullFSBasejailStorage, ZFSBasejailStorage, None]")
```

File: libiocage/lib/Jail.py [hunk 7]
```diff
@@ -143,7 +152,15 @@ def start(self):
...
 
         if self.config["basejail_type"] == "zfs":
             backend = libiocage.lib.ZFSBasejailStorage.ZFSBasejailStorage
```
Comment (igalic) [inline libiocage/lib/Jail.py]:
```
libiocage/lib/Jail.py:166: error: Incompatible types in assignment (expression has type Type[ZFSBasejailStorage], variable has type "Union[StandaloneJailStorage, NullFSBasejailStorage, ZFSBasejailStorage, None]")
```

Comment (igalic) [reply libiocage/lib/Jail.py]: for line 169, we get:

```
libiocage/lib/Jail.py:169: error: Incompatible types in assignment (expression has type Type[NullFSBasejailStorage], variable has type "Union[StandaloneJailStorage, NullFSBasejailStorage, ZFSBasejailStorage, None]")
```

and finally, for line 172, we get:

```
libiocage/lib/Jail.py:172: error: Too many arguments for "apply"
```

File: libiocage/lib/Jail.py (hunk 6 above)
Comment (gronke) [reply libiocage/lib/Jail.py]: Why is it `type Type[StandaloneJailStorage]`, but not `type "Type[Union[StandaloneJailStorage, ...."`

File: .travis/flake8.sh [hunk 8]
```diff
@@ -2,7 +2,7 @@
 # Run pep8 on all .py files in all subfolders
 
 tmpafter=$(mktemp)
-find ./libiocage/ -name \*.py -exec flake8 --ignore=E203,W391 {} + | grep -v "./libiocage/__init__.py" > ${tmpafter}
+find ./libiocage/ -name \*.py -exec flake8 --ignore=E203,E241,W391 {} + | grep -v "./libiocage/__init__.py" > ${tmpafter}
```
Comment (gronke) [inline .travis/flake8.sh]: What is `E241` good for? The documentation says `E241 (*) | multiple spaces after ‘,’`, but I can't spot why we need to ignore this rule.

Comment (skarekrow) [reply .travis/flake8.sh]: I agree. I added the ignore for spaces after `:` so we shouldn't need to add another.

Comment (igalic) [reply .travis/flake8.sh]: without this,
```
./libiocage/lib/Jail.py:72:23: E241 multiple spaces after ':'
./libiocage/lib/Jail.py:73:22: E241 multiple spaces after ':'
./libiocage/lib/Jail.py:74:23: E241 multiple spaces after ':'
./libiocage/lib/Jail.py:76:22: E241 multiple spaces after ':'
```
i've basically just followed @skarekrow's lead of how to indent… structures.

**Comment by igalic** [comment]
i'd like to close this pr.
it was a good learning experience, but it's a complete mess, especially after rebasing #37 
i've opened #54, where i'll try to do this over, and, more gradually…

**Pull request closed without merging**